File: ios/BFExplorer/Plaoc/bfs/microModule/sys/jmm/JmmNMM.swift
//
//  JmmNMM.swift
//  BFExplorer
//
//  Created by ui08 on 2023/3/6.
//

import Foundation
import Vapor

class JmmNMM: NativeMicroModule {
    var apps: [Mmid: JsMicroModule] = [:]
    
    override init() {
        super.init()
        mmid = "jmm.sys.dweb"
    }
    
    override func _bootstrap(bootstrapContext: BootStrapContext) async throws {
        routerHandler(bootstrapContext: bootstrapContext)
        
        for app in apps.values {
            bootstrapContext.dns.install(mm: app)
        }
    }
    
    private func routerHandler(bootstrapContext: BootStrapContext) {
        let installRouteHandler: RouterHandler = { request, ipc async in
            let metadataUrl = request.query[String.self, at: "metadata-url"]
            
            let jmmMetadata = await self.nativeFetch(url: metadataUrl!.decodeURIComponent()).json(JmmMetadata.self)
            self.openJmmMetadataInstallPage(jmmMetadata: jmmMetadata) { metadata in
                let jmm = JsMicroModule(metadata: metadata)
                self.apps[jmmMetadata.id] = jmm
                bootstrapContext.dns.install(mm: jmm)
            }
            
            return jmmMetadata
        }
        let uninstallRouteHandler: RouterHandler = { request, ipc async in
            let mmid = request.query[Mmid.self, at: "mmid"]
            let jmm = self.apps[mmid!]
            if jmm == nil {
                fatalError("")
            }
            self.openJmmMetadataUninstallPage(jmmMetadata: jmm!.metadata)
            return true
        }
        let queryRouteHandler: RouterHandler = { request, ipc async in
            return AppQueryResult(installedAppList: self.apps.map { $1.metadata }, installingAppList: self.installingApps.map { $1 })
        }
        let downloadRouteHandler: RouterHandler = { request, ipc async in
            let mmid = request.query[Mmid.self, at: "mmid"]
            let jmm = self.apps[mmid!]
            if self.apps[mmid!] == nil {
                let jmm = JsMicroModule(
                    metadata: JmmMetadata(
                        id: mmid!,
                        server: JmmMetadata.MainServer(
                            root: "file://bundle",
                            entry: "/cot.worker.js")))
                
                self.apps[mmid!] = jmm
            }
            
            return true
        }
        apiRouting["\(self.mmid)/install"] = installRouteHandler
        apiRouting["\(self.mmid)/uninstall"] = uninstallRouteHandler
        apiRouting["\(self.mmid)/query"] = queryRouteHandler
        apiRouting["\(self.mmid)/download"] = downloadRouteHandler
        
        // 添加路由处理方法到http路由中
        let app = HttpServer.app
        let group = app.grouped("\(mmid)")
        let httpHandler: (Request) async throws -> Response = { request async in
            await self.defineHandler(request: request)
        }
        for pathComponent in ["install", "uninstall", "query", "download"] {
            group.on(.GET, [PathComponent(stringLiteral: pathComponent)], use: httpHandler)
        }
    }
    
    struct AppQueryResult {
        let installedAppList: [JmmMetadata]
        let installingAppList: [InstallingAppInfo]
    }
    
    struct InstallingAppInfo {
        var progress: Float
        let jmmMetadata: JmmMetadata
    }
    
    private var installingApps: [Mmid: InstallingAppInfo] = [:]
    
    private func openJmmMetadataInstallPage(jmmMetadata: JmmMetadata, installDns: Callback<JmmMetadata, Void>) {
        // TODO: 下载解压压缩包
        // TODO: 使用 file://dns.sys.dweb/install 进行注册
        
    }
    
    private func openJmmMetadataUninstallPage(jmmMetadata: JmmMetadata) {
        // TODO: 未实现
    }
    
    
}

File: next/dweb-browser/src/Platforms/iOS/WebModule/Jmm/JmmNMM.cs
using DwebBrowser.MicroService;
using DwebBrowser.MicroService.Core;
using DwebBrowser.MicroService.Message;
using DwebBrowser.MicroService.Sys.Http.Net;
using DwebBrowser.Helper;
using System.Linq;

#nullable enable

namespace DwebBrowser.WebModule.Jmm;

public class JmmNMM : NativeMicroModule
{
    public override string Mmid { get; init; }

    private static Dictionary<Mmid, JsMicroModule> s_apps = new();
    public static Dictionary<Mmid, JsMicroModule> GetAndUpdateJmmNmmApps() => s_apps;

    /// <summary>
    /// 获取当前App的数据配置
    /// </summary>
    /// <param name="mmid"></param>
    /// <returns></returns>
    public static JmmMetadata? GetBfsMetaData(Mmid mmid) => s_apps.GetValueOrDefault(mmid)?.Metadata;


    public JmmNMM()
    {
        Mmid = "jmm.sys.dweb";

        Task.Run(async () =>
        {
            while (true)
            {
                await Task.Delay(1000);

                try
                {
                    await NativeFetchAsync(new Uri("file://dns.sys.dweb/open").AppendQuery("app_id", "jmm.sys.dweb".EncodeURIComponent()));
                    break;
                }
                catch
                { }
            }

            // TODO: JmmMetadataDB功能未实现
        });
    }

    protected override async Task _bootstrapAsync(IBootstrapContext bootstrapContext)
    {
        HttpRouter.AddRoute(HttpMethod.Get.Method, "/install", async (request, _) =>
        {
            var metadataUrl = request.QueryValidate<string>("metadataUrl")!;
            var jmmMetadata = await (await NativeFetchAsync(metadataUrl)).Json<JmmMetadata>()!;
            _openJmmMetadataInstallPage(jmmMetadata);

            return jmmMetadata;
        });

        HttpRouter.AddRoute(HttpMethod.Get.Method, "/uninstall", async (request, _) =>
        {
            var mmid = request.QueryValidate<string>("mmid")!;
            var jmm = s_apps.GetValueOrDefault(mmid) ?? throw new Exception("");
            _openJmmMetadataUninstallPage(jmm.Metadata);

            return true;
        });

        HttpRouter.AddRoute(HttpMethod.Get.Method, "/query", async (request, _) =>
        {
            return new AppQueryResult(
                s_apps.Values.Select(x => x.Metadata).ToList(), _installingApps.Values.ToList());
        });

        HttpRouter.AddRoute(HttpMethod.Get.Method, "/pause", async (_, ipc) =>
        {
            // TODO: 未实现JmmNMM暂停路由
            throw new NotImplementedException();
        });

        HttpRouter.AddRoute(HttpMethod.Get.Method, "/resume", async (_, ipc) =>
        {
            // TODO: 未实现JmmNMM恢复路由
            throw new NotImplementedException();
        });

        HttpRouter.AddRoute(HttpMethod.Get.Method, "/cancel", async (_, ipc) =>
        {
            // TODO: 未实现JmmNMM取消路由
            throw new NotImplementedException();
        });
    }

    private void _openJmmMetadataInstallPage(JmmMetadata jmmMetadata)
    { }

    private void _openJmmMetadataUninstallPage(JmmMetadata jmmMetadata)
    { }

    public record AppQueryResult(List<JmmMetadata> InstalledAppList, List<InstallingAppInfo> InstallingAppList);
    public record InstallingAppInfo(float Progress, JmmMetadata JmmMetadata);

    private Dictionary<Mmid, InstallingAppInfo> _installingApps = new();

    protected override async Task _onActivityAsync(IpcEvent Event, Ipc ipc)
    {
        
    }

    protected override async Task _shutdownAsync()
    {
        
    }
}

